cloudencrypt/dual.go

```diff
@@ -0,0 +1,96 @@
+package cloudencrypt
+
+import (
+	"context"
+)
+
+const (
+	mapKeyEncryptedKey   = "key"
+	mapKeyEncryptedValue = "value"
+)
+
+// DualEncryptor implements Encryptor by encrypting the value using the NativeEncryptor
+// and then encrypting the secret key using the provided encryptor.
+type DualEncryptor struct {
+	encryptor Encryptor
+}
+
+func NewDualEncryptor(ctx context.Context, encryptor Encryptor) (*DualEncryptor, error) {
+	return &DualEncryptor{
+		encryptor: encryptor,
+	}, nil
+}
+
+func (encryptor *DualEncryptor) Encrypt(ctx context.Context, value []byte, metadata ...MetadataKV) ([]byte, error) {
+	// Generate a random secret key
+	secretKey, err := generateSecretKey()
+	if err != nil {
+		return nil, err
+	}
+
+	nativeEncryptor, err := NewNativeEncryptor(secretKey)
+	if err != nil {
+		return nil, err
+	}
+
+	defer nativeEncryptor.Close()
+
+	// Encrypt the value using the random secret key
+	encryptedValue, err := nativeEncryptor.Encrypt(ctx, value)
+	if err != nil {
+		return nil, err
+	}
+
+	// Encrypt the secret key
+	encryptedSecretKey, err := encryptor.encryptor.Encrypt(ctx, secretKey, metadata...)
```

Comment:
```suggestion
	encryptedSecretKey, err := dualEncryptor.Encrypt(ctx, secretKey, metadata...)
```
If you unscope the encryptor in struct.

Reply:
No, I need to call the wrapped encryptor, this would cause recursion.